File: go/transport/rank.go
// SPDX-License-Identifier: MIT

package transport

import (
	"math"
	"sort"
)

// RankedTransport pairs a transport with its composite score.
type RankedTransport struct {
	// Transport is the ranked transport backend.
	Transport TransportService
	// Score is the composite score (higher = better).
	Score float64
}

// Rank scores and sorts a slice of transports in descending composite-score
// order.  Unavailable transports are excluded from the result.
//
// Transports that return nil from Metrics() are ranked by their declared
// MaxBandwidthBps and PowerCostRelative alone (conservative fallback).
//
// Returns nil (not an empty slice) when no transport passes the filter.
func Rank(transports []TransportService) []RankedTransport {
	if len(transports) == 0 {
		return nil
	}

	result := make([]RankedTransport, 0, len(transports))

	for _, t := range transports {
		if !t.IsAvailable() {
			continue
		}

		var score float64
		if m := t.Metrics(); m != nil {
			score = m.CompositeScore(t.MaxBandwidthBps(), t.PowerCostRelative())
		} else {
			// Fallback: no EWMA data — score by declared bandwidth / power.
			power := math.Max(float64(t.PowerCostRelative()), 1.0)
			score = float64(t.MaxBandwidthBps()) * 0.1 * 0.95 / power
		}

		result = append(result, RankedTransport{
			Transport: t,
			Score:     score,
		})
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].Score > result[j].Score
	})

	return result
}

File: python/aethernet/transport/rank.py
# SPDX-License-Identifier: MIT

"""Transport ranking — orders available transports by composite score."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, List

if TYPE_CHECKING:
    from aethernet.transport.transport_service import TransportService


@dataclass
class RankedTransport:
    """A transport paired with its pre-computed composite score."""

    transport: "TransportService"
    score: float


def rank_transports(transports: List["TransportService"]) -> List[RankedTransport]:
    """
    Order available transports by composite score (highest first).

    Unavailable transports are excluded.  For transports without live metrics,
    the static prior (max_bandwidth_bps / power_cost_relative) is used.

    Args:
        transports: All registered transport backends.

    Returns:
        List of :class:`RankedTransport` sorted descending by score.
    """
    result: List[RankedTransport] = []

    for t in transports:
        if not t.is_available:
            continue

        m = t.metrics
        if m is not None:
            score = m.composite_score(t.max_bandwidth_bps, t.power_cost_relative)
        else:
            power = max(t.power_cost_relative, 1)
            score = t.max_bandwidth_bps / power

        result.append(RankedTransport(transport=t, score=score))

    result.sort(key=lambda rt: rt.score, reverse=True)
    return result
